staging/src/k8s.io/apiserver/pkg/storage/testing/store_tests.go

```diff
@@ -1048,77 +1049,86 @@ func seedMultiLevelData(ctx context.Context, store storage.Interface) (string, [
 }
 
 func RunTestGetListNonRecursive(ctx context.Context, t *testing.T, store storage.Interface) {
-	prevKey, prevStoredObj := TestPropagateStore(ctx, t, store, &example.Pod{ObjectMeta: metav1.ObjectMeta{Name: "prev"}})
-
+	key, prevStoredObj := testPropagateStore(ctx, t, store, &example.Pod{ObjectMeta: metav1.ObjectMeta{Name: "foo"}})
 	prevRV, _ := strconv.Atoi(prevStoredObj.ResourceVersion)
-	key, storedObj := TestPropagateStore(ctx, t, store, &example.Pod{ObjectMeta: metav1.ObjectMeta{Name: "foo"}})
+
+	storedObj := &example.Pod{}
+	if err := store.GuaranteedUpdate(ctx, key, storedObj, false, nil,
+		func(_ runtime.Object, _ storage.ResponseMeta) (runtime.Object, *uint64, error) {
+			newPod := prevStoredObj.DeepCopy()
+			newPod.Annotations = map[string]string{"version": "second"}
+			return newPod, nil, nil
+		}, nil); err != nil {
+		t.Fatalf("update failed: %v", err)
+	}
 	currentRV, _ := strconv.Atoi(storedObj.ResourceVersion)
 
 	tests := []struct {
-		name             string
-		key              string
-		pred             storage.SelectionPredicate
-		expectedOut      []*example.Pod
-		rv               string
-		rvMatch          metav1.ResourceVersionMatch
-		expectRVTooLarge bool
+		name                 string
+		key                  string
+		pred                 storage.SelectionPredicate
+		expectedOut          []example.Pod
+		expectedAlternatives [][]example.Pod
+		rv                   string
+		rvMatch              metav1.ResourceVersionMatch
+		expectRVTooLarge     bool
 	}{{
 		name:        "existing key",
 		key:         key,
 		pred:        storage.Everything,
-		expectedOut: []*example.Pod{storedObj},
+		expectedOut: []example.Pod{*storedObj},
 	}, {
-		name:        "existing key, resourceVersion=0",
-		key:         key,
-		pred:        storage.Everything,
-		expectedOut: []*example.Pod{storedObj},
-		rv:          "0",
+		name:                 "existing key, resourceVersion=0",
```

Comment:
I shuffled the order of tests to force better testing of watchcache.

Basically when we were testing "resourceVersion=current" before "resourceVersion=previous", watchcache was synchronizing to "current" on the first testcase and was never on "previous" in the second one.
Changing the order means that for "previous" testcase it can actually be on both.

Reply:
Can we randomize the keys used for the test, make each test case `t.Parallel()` and independent, and not allow them to influence each other?

Reply:
It's not what mattered here. What matter is that watchcache is not synchronous.

And I think for the semantic "NotOlderThan" it actually is fairly useful to test that semantic - for which the answer is not unique (if there were N changes after a given RV, all results are "correct" ones).

So I think that what we have currently is actually desired from the POV of testing the semantic of the interface.

Reply:
Sure, but it's implicit - I would walk in there and break it without having any idea! If we parallelized it and made cache hits from different tests independent, we could further write a test that's specific to the cache that does a number of calls in series, as the generic store interface doesn't provide any such guarantees, right?

Reply:
ohh - so you're effectively asking form actually running test cases in parallel? That sounds reasonable to me.

I was trying to make that happen by using t.Parallel() but I don't know how to make that work nicely...
As described in https://engineering.mercari.com/en/blog/entry/20220408-how_to_use_t_parallel/
t.Parallel() (for test subcases using t.Parallel()) are actually paused until the main test case finishes.
That makes the cleanup harder to run...

We can switch to t.Cleanup() in function calling those (from defer) which seems to work.
I'm not sure it's the nicest thing, but work.

I added that in a separate commit for now for you to take a look (and then even with the original order I can reproduce the problem of forgetting to add `prevStoredObj` as a correct result in
"existing key, resourceVersion=previous, resourceVersionMatch=notOlderThan" testcase.
[within 10 runs of tests]
Which is nice.

See the third commit if that's what you meant.